File: feriwala_customer/lib/services/appwrite_otp_service.dart
import 'dart:convert';
import 'package:http/http.dart' as http;
import '../config/app_config.dart';

/// SMS OTP service that routes through our backend.
/// Backend uses Appwrite server API key to send/verify OTP.
class AppwriteOtpService {
  AppwriteOtpService._();
  static final AppwriteOtpService instance = AppwriteOtpService._();

  String? _userId;
  String? get lastUserId => _userId;

  /// Send OTP to phone number via our backend.
  /// [phone] must be in E.164 format e.g. +919876543210
  Future<bool> sendOtp(String phone) async {
    try {
      final response = await http.post(
        Uri.parse('${AppConfig.apiBaseUrl}/auth/send-otp'),
        headers: {'Content-Type': 'application/json'},
        body: jsonEncode({'phone': phone}),
      );

      if (response.statusCode >= 200 && response.statusCode < 300) {
        final data = jsonDecode(response.body) as Map<String, dynamic>;
        _userId = data['data']?['userId'];
        return data['success'] == true;
      }
      return false;
    } catch (_) {
      return false;
    }
  }

  /// Verify OTP code via our backend.
  Future<bool> verifyOtp(String otp) async {
    if (_userId == null) return false;
    try {
      final response = await http.post(
        Uri.parse('${AppConfig.apiBaseUrl}/auth/verify-otp'),
        headers: {'Content-Type': 'application/json'},
        body: jsonEncode({'userId': _userId, 'otp': otp}),
      );

      if (response.statusCode >= 200 && response.statusCode < 300) {
        final data = jsonDecode(response.body) as Map<String, dynamic>;
        return data['success'] == true;
      }
      return false;
    } catch (_) {
      return false;
    }
  }
}

File: backend/src/services/appwriteOtpService.js
const axios = require('axios');

const APPWRITE_ENDPOINT = process.env.APPWRITE_ENDPOINT || 'https://tor.cloud.appwrite.io/v1';
const APPWRITE_PROJECT_ID = process.env.APPWRITE_PROJECT_ID || '6a022ea60030dfa3545e';
const APPWRITE_API_KEY = process.env.APPWRITE_API_KEY || '';

const headers = {
  'Content-Type': 'application/json',
  'X-Appwrite-Project': APPWRITE_PROJECT_ID,
  'X-Appwrite-Key': APPWRITE_API_KEY,
};

/**
 * Send SMS OTP to phone number via Appwrite Phone auth.
 * @param {string} phone - E.164 format e.g. +919876543210
 * @returns {{ success: boolean, userId?: string, error?: string }}
 */
async function sendPhoneOtp(phone) {
  try {
    const res = await axios.post(
      `${APPWRITE_ENDPOINT}/users/create-token`,
      { userId: 'unique()', phone },
      { headers }
    );
    return { success: true, userId: res.data.userId || res.data.$id };
  } catch (error) {
    // Fallback: try the account tokens endpoint (client-style)
    try {
      const res2 = await axios.post(
        `${APPWRITE_ENDPOINT}/account/tokens/phone`,
        { userId: 'unique()', phone },
        { headers: { 'Content-Type': 'application/json', 'X-Appwrite-Project': APPWRITE_PROJECT_ID } }
      );
      return { success: true, userId: res2.data.userId || res2.data.$id };
    } catch (err2) {
      const msg = err2.response?.data?.message || error.response?.data?.message || error.message;
      return { success: false, error: msg };
    }
  }
}

/**
 * Verify SMS OTP code.
 * @param {string} userId - User ID from sendPhoneOtp
 * @param {string} otp - 6-digit code
 * @returns {{ success: boolean, error?: string }}
 */
async function verifyPhoneOtp(userId, otp) {
  try {
    await axios.put(
      `${APPWRITE_ENDPOINT}/account/sessions/phone`,
      { userId, secret: otp },
      { headers: { 'Content-Type': 'application/json', 'X-Appwrite-Project': APPWRITE_PROJECT_ID } }
    );
    return { success: true };
  } catch (error) {
    const msg = error.response?.data?.message || error.message;
    return { success: false, error: msg };
  }
}

module.exports = { sendPhoneOtp, verifyPhoneOtp };
